File: KIJUNG/PRO_WEEK_06/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    sort(dungeons.begin(), dungeons.end());

    do{
        int cnt = 0;
        int need = k;
        for(vector<int> dungeon: dungeons){
            if(need<dungeon[0]) break;
            else{
                need-=dungeon[1];
                cnt++;
            }
        }
        answer=max(cnt, answer);
    }while(next_permutation(dungeons.begin(), dungeons.end()));

    return answer;
}

File: DONGUK/프로그래머스_WEEK 7/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = 0
    
    N = len(dungeons)
    
    orders = [i for i in range(1, N+1)]
    
    perm_orders = list(permutations(orders, N))
    
    for order in perm_orders:
        now_hp = k
        cnt = 0
        for ind in order:
            req_hp, lose_hp = dungeons[ind-1][0], dungeons[ind-1][1]
            
            if now_hp >= req_hp:
                cnt += 1
                now_hp -= lose_hp
                
            else:
                break
                
        answer = max(answer, cnt)
        
    return answer